### PR TITLE
[Prototyping] load linter config from repository

### PR DESCRIPTION
**Pull request by MissingRoberto**
- Fetch linter config from repository on linting.
- Use a factory pattern to create linters.

# Remarks

- Excluded linting in parser for now due to package intricacies. I will revisit this later.
- I believe that the linter config must live in the repository as people want to run linters locally before push. Happy to discuss that.

### CODE REVIEW

File: pkg/registry/apis/provisioning/lint/dashboard.go [hunk 1]
```diff
@@ -8,10 +8,35 @@ import (
 
 	"github.com/grafana/dashboard-linter/lint"
 	provisioning "github.com/grafana/grafana/pkg/apis/provisioning/v0alpha1"
+	"gopkg.in/yaml.v2"
```
Comment (ryantxu) [inline pkg/registry/apis/provisioning/lint/dashboard.go:11]: could/should we depend on v3 or is that from the linter package?

File: public/app/plugins/panel/logs/panelcfg.cue
Comment (ryantxu) [inline public/app/plugins/panel/logs/panelcfg.cue:1]: unrelated change

Comment (MissingRoberto) [reply public/app/plugins/panel/logs/panelcfg.cue:1]: I tried to revert the change but unsuccessfully 😆

File: pkg/registry/apis/provisioning/lint/dashboard.go (hunk 1 above)
Comment (MissingRoberto) [reply pkg/registry/apis/provisioning/lint/dashboard.go:11]: the import was added automatically. I will use `v3`.

File: public/app/plugins/panel/logs/panelcfg.cue
Comment (MissingRoberto) [reply public/app/plugins/panel/logs/panelcfg.cue:1]: image
It's a pre-commit hook.

File: pkg/registry/apis/provisioning/lint/lint.go [hunk 2]
```diff
@@ -9,3 +9,9 @@ import (
 type Linter interface {
 	Lint(ctx context.Context, data []byte) ([]provisioning.LintIssue, error)
 }
+
+type LinterFactory interface {
+	ConfigPath() string
+	NewFromConfig(cfg []byte) (Linter, error)
```
[the remark this answers is not in the document]
Comment (MissingRoberto) [reply pkg/registry/apis/provisioning/lint/lint.go:15]: @ryantxu  In multi-tenant land we also have access to the repository to fetch this from the repository as we fetch any other file. So, I am not sure which part concerns you there :) 

The thing is that if we put it as part of the spec, people won't be able to run it locally to fix things there in the repository. That's why you always find linter config as part of the repository with dotfiles or other config files.

[the remark this answers is not in the document]
Comment (ryantxu) [reply pkg/registry/apis/provisioning/lint/lint.go:15]: is `ConfigPath() string` the path to the file *within* the remote repository?  EG, you save the lint rules yaml as a file in git?

Interesting

File: pkg/registry/apis/provisioning/repository/github.go [hunk 3]
```diff
@@ -25,12 +25,12 @@ import (
...
-	config   *provisioning.Repository
-	gh       pgh.Client
-	baseURL  *url.URL
-	linter   lint.Linter
-	renderer PreviewRenderer
+	logger        *slog.Logger
+	config        *provisioning.Repository
+	gh            pgh.Client
+	baseURL       *url.URL
+	linterFactory lint.LinterFactory
```
Comment (ryantxu) [inline pkg/registry/apis/provisioning/repository/github.go:32]: Maybe a different issue... should githubRepository be using the parser to validate?  This would give it access to the same logic we can use everywhere that also checks what happens with dryRun, etc

File: pkg/registry/apis/provisioning/resources/parser.go [hunk 4]
```diff
@@ -148,6 +148,7 @@ func (r *FileParser) Parse(ctx context.Context, logger *slog.Logger, info *repos
 			}
 		}
 		linter := lint.NewDashboardLinter()
+		// TODO: pass the config file
```
Comment (ryantxu) [inline pkg/registry/apis/provisioning/resources/parser.go:151]: Loading lint from the repo is interesting, but I think the path should be saved in the repo spec, not auto-detected

**Review by ryantxu — APPROVED** [review]
Changes to linter seem good/progress

We need to rename "parser" to something more appropriate, and use it everywhere we are trying to validate (but that is a new issue!)

**Comment by MissingRoberto** [comment]
@ryantxu yes, I agree. The reason why I didn't add the required changes to the parser was because I struggled with the current package structure and how the different how the structs / abstractions depend on each other :)
